File: 数学题/134.java
// https://leetcode-cn.com/problems/gas-station/solution/jia-you-zhan-by-leetcode-solution/
// 可以推出：若从x可以走到y，但不能走到(y+1)，则x~y之间的任意一个加油站z都不能走到(y+1)
// 因此x走到(y+1)失败时，下次可以直接从(y+1)开始尝试
// 时间复杂度O(n)。暴力解法的复杂度O(n^2)
class Solution {
    public int canCompleteCircuit(int[] gas, int[] cost) {
        int n = gas.length;
        if (n == 0) {
            return -1;
        }
        int start = 0;   // 开始位置
        while (start < n) {
            int cnt = 0;  // 向后走几个加油站
            int sumGas = 0, sumCost = 0;
            while (cnt < n) {
                int j = (start + cnt) % n;  // 尝试走到的加油站
                sumCost += cost[j];  // 需要消耗的
                sumGas += gas[j];   // 能够得到的
                if (sumGas < sumCost) {   // 不能走到j
                    break;
                } else {    // 可以走到j，再往后走一个
                    ++cnt;
                }
            }
            // 判断是由于break退出的循环，还是自退出的循环
            if (cnt == n) {
                return start;
            }
            start = start + cnt + 1;
        }
        return -1;
    }
}

File: 模拟/134.go
// 暴力解法复杂度O(n^2)，不行
// 剪枝策略：当i走不到j时，i之后的点也走不到j，因此下一次尝试的起点直接设置为j
// 证明：假设油量为have，i 走到 j 时have < 0，说明从 i 走到 (i+1)~(j-1) 之间的位置have > 0。若设置 (i+1)~(j-1) 之间的某一位置做起点，这个位置上的have == 0，更不可能走到 j
func canCompleteCircuit(gas []int, cost []int) int {
	sum := 0
	for i := range gas {
		sum += gas[i] - cost[i]
	}
	if sum < 0 {
		return -1
	}
  // 起点是0
	have := gas[0]
	start := 0
	for i := 1; i < len(gas); i++ {
		have -= cost[i - 1]
    // 走不到 i
		if have < 0 {
			start = i   // 起点设为 i
			have = gas[i]
		} else {  // 获得位置 i 的油
			have += gas[i]
		}
	}
	if start < len(gas) {
		return start
	}
	return -1
}
